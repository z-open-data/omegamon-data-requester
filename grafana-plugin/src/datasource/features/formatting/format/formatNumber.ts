import { ColumnMetadata } from 'public-domain';

import { convertToNumber } from './convertToNumber';

function scale(value: number, scaleFactor: number | undefined): number {
  if (scaleFactor == null) {
    return value;
  }
  const formattedValue = value / 10 ** scaleFactor;
  return formattedValue;
}

/**
 * Formats number according to PRINTF and scaleFactor in supplied column metadata
 * @param value - decimal number or hex number (e.g. `10` or `"0XA"` or `"A"`)
 * @returns decimal number (as number) or hex number (as string)
 */
export function formatNumber(value: number | string, columnMd: ColumnMetadata): string | number {
  const asNumber = convertToNumber(value);
  return scale(asNumber, columnMd.scaleFactor);
}
