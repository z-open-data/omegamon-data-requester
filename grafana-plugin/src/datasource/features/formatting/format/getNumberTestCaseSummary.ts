import { ColumnMetadata } from 'public-domain';

export function getNumberTestCaseSummary(val: string | number, columnMd: ColumnMetadata): string {
  const scaleFactorSummary = columnMd.scaleFactor ? `scaleFactor=${columnMd.scaleFactor}` : '';
  const printfSummary = columnMd.printf ? `PRINTF="${columnMd.printf}"` : '';
  const valueFormattedForPrinting = typeof val === 'string' ? `"${val}"` : val;
  const valueSummary = `val=${valueFormattedForPrinting}`;
  return `${valueSummary} ${printfSummary} ${scaleFactorSummary}`;
}
