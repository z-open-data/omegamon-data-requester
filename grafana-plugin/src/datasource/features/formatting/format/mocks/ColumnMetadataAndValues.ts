import { ColumnMetadata } from 'public-domain';

export interface ColumnMetadataAndValues<TValue = string | number> {
  values: readonly TValue[];
  columnMetadata: ColumnMetadata;
}
