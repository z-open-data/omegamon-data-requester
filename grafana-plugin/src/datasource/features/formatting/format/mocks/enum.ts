import { EnumMap, ColumnMetadata } from 'public-domain';

export const stringEnumColumnMetadataAndValues = {
  columnMetadata: {
    attributeName: 'Thread\\Type',
    defaultSortDirection: 'A',
    description: 'The thread type.',
    enums: {
      '0': 'Unknown',
      '1': 'DB Access',
      '3': 'SYSTEM',
      '5': 'UTILITY',
      '6': 'RRSAF',
      '7': 'TSO',
      '9': 'Allied',
      A: 'IMS',
      B: 'BATCH',
      C: 'CICS',
      'IN-TRIGGER': 'IN-TRIGGER',
      'IN-USER-FUNC': 'IN-USER-FUNC',
      INDOUBT: 'IN-DOUBT',
    } as EnumMap,
    extensions: {
      MEMBER: 'Thread_Type',
      NULL: 'N',
      PARM: "H'14',H'00',H'0001',C'C31 ',F'0192'",
      SHARE: 'Y',
      STRUCT: 'ALL_THREADS',
      TEXT: 'THREAD TYPE',
    },
    id: 'TDIDTTYP',
    maxLength: 1,
    name: 'Thread\nType',
    odpName: '',
    sortType: 'lexical',
    type: 'string',
    version: 0,
    atomize: false,
  } as ColumnMetadata,
  values: ['0', '3', 'A', 'IN-TRIGGER', 'IN-USER-FUNC', 'INDOUBT'],
};

export const numberEnumColumnMetadataAndValues = {
  columnMetadata: {
    attributeName: 'Performance\\Index',
    defaultSortDirection: 'A',
    description: 'This field is related to the z/OS Workload manager.',
    enums: {
      '0': 'N/A',
      '123': 'some other enum',
      '-123': 'even other enum',
    } as EnumMap,
    extensions: {
      MEMBER: 'Performance_Index',
      NULL: 'N',
      PARM: "H'14',H'00',H'0004',C'C31 ',F'0416'",
      RANGE: '1-999',
      SHARE: 'Y',
      STRUCT: 'ALL_THREADS',
      TEXT: 'PERFORMANCE INDEX THIS PERIOD',
    },
    id: 'ENCLPIDX',
    maxLength: 4,
    name: 'Performance\nIndex',
    odpName: '',
    sortType: 'lexical',
    type: 'number',
    version: 0,
    atomize: false,
  } as ColumnMetadata,
  values: [0, 123, -123],
};
