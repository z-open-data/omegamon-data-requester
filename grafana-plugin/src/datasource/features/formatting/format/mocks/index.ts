export { ColumnMetadataAndValues } from './ColumnMetadataAndValues';
export { numberEnumColumnMetadataAndValues, stringEnumColumnMetadataAndValues } from './enum';
export {
  columnMdsWithPrintf,
  columnMdsWithoutHexPrintf,
  numberCasesAsNumbersAndAsStrings,
  numberColumnMds,
  brokenNumberCasesAsStrings,
  numberCasesAsNumbers,
} from './number';
export { stringColumnMetadata, stringColumnValues } from './string';
export { invalidItmTimestamps, itmTimestampTestValues, timestampColumnMd, validItmTimestamps } from './timestamp';
