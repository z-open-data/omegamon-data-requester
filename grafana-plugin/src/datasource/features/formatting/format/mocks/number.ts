import { ColumnMetadata } from 'public-domain';

export const numberCasesAsNumbers = [-123, -1, 0, 1, 10, 123, 72491];
export const numberCasesAsStrings = ['0', '000123', '0xA', '0X01F'];
export const numberCasesAsNumbersAndAsStrings = [...numberCasesAsNumbers, ...numberCasesAsStrings];
export const brokenNumberCasesAsStrings = [
  'hello',
  'O',
  'Ox0',
  '0xO',
  '-seven',
  '0xABSENCE',
  '0xABCDEFG',
  ',.;[]',
  ';.123',
  '0x<>',
];

const columnMdWithUnits = {
  attributeName: 'User\\Process\\Time',
  defaultSortDirection: 'A',
  description:
    'The total user CPU time, in seconds and microseconds, used by the database manager agent process, the unit of work, or the statement.',
  extensions: {
    NULL: 'N',
    OPTION: 'SORTBYINTERNAL',
    PARM: "H'14',H'00',H'0004',C'C31 ',F'0284'",
    SHARE: 'Y',
    UNITS: 'milliseconds',
    STRUCT: 'DB2_CONNECT_TASKLIST',
    MEMBER: 'EESTL004',
    RANGE: '1-99999999',
  },
  id: 'EESTL004',
  maxLength: 4,
  name: 'User\nProcess\nTime',
  odpName: '',
  sortType: 'lexical',
  type: 'number',
  version: 0,
  unit: 'milliseconds',
  atomize: false,
} satisfies ColumnMetadata;

const columnMdWithoutPrintfWithoutScaleFactor = {
  attributeName: 'Uniqueness\\Value2',
  defaultSortDirection: 'A',
  description: 'The DB2 thread uniqueness value.',
  extensions: {
    MEMBER: 'Uniqueness_Value2',
    NULL: 'N',
    PARM: "H'14',H'00',H'0002',C'C31 ',F'0428'",
    RANGE: '1-999',
    SHARE: 'Y',
    STRUCT: 'ALL_THREADS',
    TEXT: 'UNIQUENESS VALUE',
  },
  id: 'QWHSLUUV2',
  maxLength: 2,
  name: 'Uniqueness\nValue2',
  odpName: '',
  sortType: 'lexical',
  type: 'number',
  version: 0,
  atomize: false,
} satisfies ColumnMetadata;

const columnMdWithoutPrintfWithScaleFactor = {
  attributeName: 'Read_Hit\\Percent',
  defaultSortDirection: 'A',
  description: 'The percentage of READ requests that were satisfied from cache.',
  scaleFactor: 2,
  extensions: {},
  id: 'RD_TOT_HP',
  maxLength: 4,
  name: 'Read Hit\nPercent',
  odpName: '',
  sortType: 'lexical',
  type: 'number',
  version: 0,
  atomize: false,
} satisfies ColumnMetadata;

const columnMdWithNonHexPrintfWithScaleFactor = {
  attributeName: 'Read_Hit\\Percent',
  defaultSortDirection: 'A',
  description: 'The percentage of READ requests that were satisfied from cache.',
  scaleFactor: 1,
  printf: '"%5.1f"',
  extensions: {
    RANGE: '0-1000',
    SHARE: 'Y',
    UNITS: '".1 %"',
    USAGE: 'N',
  },
  id: 'RD_TOT_HP',
  maxLength: 4,
  name: 'Read Hit\nPercent',
  odpName: '',
  sortType: 'lexical',
  type: 'number',
  version: 0,
  atomize: false,
} satisfies ColumnMetadata;

/**
 * Possible cases (at least in theory):
 * 1. no printf, no scaleFactor
 * 2. nonHexPrintf, no scaleFactor
 * 3. no printf, scaleFactor
 * 4. nonHexPrintf, scaleFactor
 */

export const columnMdsWithoutHexPrintf: ColumnMetadata[] = [
  columnMdWithoutPrintfWithScaleFactor,
  columnMdWithoutPrintfWithoutScaleFactor,
  columnMdWithNonHexPrintfWithScaleFactor,
  columnMdWithUnits,
];

export const columnMdsWithPrintf: ColumnMetadata[] = [columnMdWithNonHexPrintfWithScaleFactor];

export const numberColumnMds: ColumnMetadata[] = [...columnMdsWithoutHexPrintf];
