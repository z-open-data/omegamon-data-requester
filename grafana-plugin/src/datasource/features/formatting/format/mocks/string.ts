import { ColumnMetadata } from 'public-domain';

export const stringColumnValues = [
  'RSD1',
  'VTAM:RSD2',
  'IB1D:RSD4:DB2',
  'ITE2OI0',
  'TSO',
  'DD66C840FFEF8E74',
  '!IB1D CANCEL THREAD (000010)',
];

export const stringColumnMetadata = {
  attributeName: 'Cancel_Command',
  defaultSortDirection: 'A',
  description: 'The command string needed to cancel a thread.',
  extensions: {
    DVERSION: '4',
    NULL: 'N',
    PARM: "H'14',H'00',H'0080',C'C31 ',F'1128'",
    SHARE: 'Y',
    STRUCT: 'ALL_THREADS',
  },
  id: 'KILLTHREAD',
  maxLength: 80,
  name: 'Cancel\nCommand',
  odpName: '',
  sortType: 'lexical',
  type: 'string',
  version: 0,
  atomize: false,
} satisfies ColumnMetadata;
