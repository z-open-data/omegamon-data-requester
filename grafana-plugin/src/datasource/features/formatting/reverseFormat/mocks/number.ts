export const negativeNumbersAsHex2sComplements = [
  '0xFFFFFFFF',
  '0xFFFFBA98',
  '0xA1234567',
  '0x80000000',
  '0x8A3A1234',
  '0X000ffffffff', // X uppercase & prepended with 0s & lowercase hex digit letters
];

export const nonNegativeNumbersAsHex = [
  '0x0',
  '0x1',
  '0xA',
  '0x000C', // prepended with 0s
  '0xABCDEF',
  '0x1234567',
  '0X0', // X uppercase
  '0X0A', // X uppercase & prepended with 0s
  '0X00abc', // X uppercase & prepended with 0s & lowercase hex digit letters
];

export const hexNumbers = [...negativeNumbersAsHex2sComplements, ...nonNegativeNumbersAsHex];

export const stringifiedDecimalNumbers = [
  '-123',
  '-1',
  '0',
  '1',
  '10',
  '123',
  '14834921',
  '0.1',
  '0.00234',
  '-0.723',
  '000123', // prepended with 0s
  '  123  ', // untrimmed
  '  -0.123  ', // untrimmed
];

export const simpleNumbers = [-123, -1, 0, 1, 10, 123, 14834921, 0.1, 0.00234, -0.723];

export const simpleAndHexNumbers = [...simpleNumbers, ...hexNumbers, ...stringifiedDecimalNumbers];
