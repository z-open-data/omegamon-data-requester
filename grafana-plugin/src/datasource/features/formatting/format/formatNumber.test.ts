import { formatNumber } from './formatNumber';
import { getNumberTestCaseSummary } from './getNumberTestCaseSummary';
import { numberColumnMds, numberCasesAsNumbersAndAsStrings, brokenNumberCasesAsStrings } from './mocks';

describe(formatNumber.name, () => {
  numberColumnMds.forEach((columnMd) => {
    numberCasesAsNumbersAndAsStrings.forEach((val) => {
      const testCaseSummary = getNumberTestCaseSummary(val, columnMd);
      test(testCaseSummary, () => {
        expect(formatNumber(val, columnMd)).toMatchSnapshot();
      });
    });

    brokenNumberCasesAsStrings.forEach((val) => {
      test(`throws error when value "${val}" cannot be converted to number`, () => {
        expect(() => {
          formatNumber(val, columnMd);
        }).toThrow();
      });
    });
  });
});
