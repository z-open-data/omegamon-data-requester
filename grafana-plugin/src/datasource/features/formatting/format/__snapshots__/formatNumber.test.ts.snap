// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatNumber val="0"   1`] = `0`;

exports[`formatNumber val="0"   2`] = `0`;

exports[`formatNumber val="0"  scaleFactor=2 1`] = `0`;

exports[`formatNumber val="0" PRINTF=""%5.1f"" scaleFactor=1 1`] = `0`;

exports[`formatNumber val="000123"   1`] = `123`;

exports[`formatNumber val="000123"   2`] = `123`;

exports[`formatNumber val="000123"  scaleFactor=2 1`] = `1.23`;

exports[`formatNumber val="000123" PRINTF=""%5.1f"" scaleFactor=1 1`] = `12.3`;

exports[`formatNumber val="0X01F"   1`] = `31`;

exports[`formatNumber val="0X01F"   2`] = `31`;

exports[`formatNumber val="0X01F"  scaleFactor=2 1`] = `0.31`;

exports[`formatNumber val="0X01F" PRINTF=""%5.1f"" scaleFactor=1 1`] = `3.1`;

exports[`formatNumber val="0xA"   1`] = `10`;

exports[`formatNumber val="0xA"   2`] = `10`;

exports[`formatNumber val="0xA"  scaleFactor=2 1`] = `0.1`;

exports[`formatNumber val="0xA" PRINTF=""%5.1f"" scaleFactor=1 1`] = `1`;

exports[`formatNumber val=-1   1`] = `-1`;

exports[`formatNumber val=-1   2`] = `-1`;

exports[`formatNumber val=-1  scaleFactor=2 1`] = `-0.01`;

exports[`formatNumber val=-1 PRINTF=""%5.1f"" scaleFactor=1 1`] = `-0.1`;

exports[`formatNumber val=-123   1`] = `-123`;

exports[`formatNumber val=-123   2`] = `-123`;

exports[`formatNumber val=-123  scaleFactor=2 1`] = `-1.23`;

exports[`formatNumber val=-123 PRINTF=""%5.1f"" scaleFactor=1 1`] = `-12.3`;

exports[`formatNumber val=0   1`] = `0`;

exports[`formatNumber val=0   2`] = `0`;

exports[`formatNumber val=0  scaleFactor=2 1`] = `0`;

exports[`formatNumber val=0 PRINTF=""%5.1f"" scaleFactor=1 1`] = `0`;

exports[`formatNumber val=1   1`] = `1`;

exports[`formatNumber val=1   2`] = `1`;

exports[`formatNumber val=1  scaleFactor=2 1`] = `0.01`;

exports[`formatNumber val=1 PRINTF=""%5.1f"" scaleFactor=1 1`] = `0.1`;

exports[`formatNumber val=10   1`] = `10`;

exports[`formatNumber val=10   2`] = `10`;

exports[`formatNumber val=10  scaleFactor=2 1`] = `0.1`;

exports[`formatNumber val=10 PRINTF=""%5.1f"" scaleFactor=1 1`] = `1`;

exports[`formatNumber val=123   1`] = `123`;

exports[`formatNumber val=123   2`] = `123`;

exports[`formatNumber val=123  scaleFactor=2 1`] = `1.23`;

exports[`formatNumber val=123 PRINTF=""%5.1f"" scaleFactor=1 1`] = `12.3`;

exports[`formatNumber val=72491   1`] = `72491`;

exports[`formatNumber val=72491   2`] = `72491`;

exports[`formatNumber val=72491  scaleFactor=2 1`] = `724.91`;

exports[`formatNumber val=72491 PRINTF=""%5.1f"" scaleFactor=1 1`] = `7249.1`;
